# pangraph tutorial

This repository contains code for a pangraph tutorial, prepared in occasion of the [
Applied Bioinformatics & Public Health Microbiology](https://coursesandconferences.wellcomeconnectingscience.org/event/applied-bioinformatics-public-health-microbiology-20230503/) conference.

This tutorial consists in using pangraph to study some properties of a dataset of *E. coli* ST131 isolates downloaded from RefSeq.

## setup

The tutorial can be executed using `snakemake`. Running the code requires:
- a working `conda` and `mamba` installation
- `snakemake`
- having the `pangraph` binary available in your path

All the files can then be generated by running:
```bash
snakemake --profile local all
```

## analysis overview

The analysis is divided in three parts:

1. examining the structural diversity around bla CTX-M-15 genes.
2. building a pangenome graph for a part of the dataset and looking at difference in private sequence compared to core genome divergence.
3. marginalizing the graph on a pair of isolates and visualizing the pairwise projection.

Each part has a corresponding set of notes in the `notes` folder.